[AutoDiff] Simplify substitution map logic.

dan-zheng opened the pull request:
Simplify `PrimalGenCloner::getOrDiagnoseAssociatedFunctionSubstitutionMap` using
`SubsMap` member and `SubstitutionMap::get` constructor.

dan-zheng committed:
[AutoDiff] Simplify substitution map logic.

Simplify `PrimalGenCloner::getOrDiagnoseAssociatedFunctionSubstitutionMap` using
`SubsMap` member and `SubstitutionMap::get` constructor.

rxwei commented:
@swift-ci please test tensorflow

slavapestov commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2506,8 +2489,10 @@ bool PrimalGen::performSynthesis(FunctionSynthesisItem item) {
-    substMap = substMap.subst(primalSubstMap);
+    substMap = SubstitutionMap::get(
+        primalGenEnv->getGenericSignature(),
+        QuerySubstitutionMap{primalGenEnv->getForwardingSubstitutionMap()},
```
substMap is equivalent to just getForwardingSubstitutionMap() here

slavapestov replied:
I mean primalGenEnv->getForwardingSubstitutionMap()

dan-zheng replied:
Thanks! Logic revamped in https://github.com/apple/swift/pull/22243/commits/0fb115dcc725e543b78d4994ace58eabf8ed69cf. (The old logic was simpler but led to stdlib compilation crash.)

slavapestov commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2314,18 +2311,12 @@ class PrimalGenCloner final
-      // it is valid. Continue.
+      // If requirement type has no type parameters and is not an archetype
+      // type, it is valid. Continue.
       if (!reqType->isTypeParameter() && !reqType->hasArchetype())
```
hasArchetype() will always be false here

dan-zheng replied:
Logic revamped in https://github.com/apple/swift/pull/22243/commits/0fb115dcc725e543b78d4994ace58eabf8ed69cf.

dan-zheng committed:
Fix substitution map logic.

The previous logic was simpler but incorrect (caused stdlib compilation failure).
Add tests for more coverage.

dan-zheng commented:
I believe stdlib compilation failures are fixed, ready for re-review.
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
Merging. I'm happy to address comments later.
